metecho/api/serializers.py

```diff
@@ -361,26 +361,53 @@ def create(self, validated_data):
     def update(self, instance, validated_data):
         user = getattr(self.context.get("request"), "user", None)
         originating_user_id = str(user.id) if user else None
-        if instance.assigned_dev != validated_data["assigned_dev"]:
-            if validated_data.get("should_alert_dev"):
-                self.try_send_assignment_emails(instance, "dev", validated_data, user)
-            orgs = instance.scratchorg_set.active().filter(
-                org_type=SCRATCH_ORG_TYPES.Dev
-            )
-            for org in orgs:
-                org.queue_delete(originating_user_id=originating_user_id)
-        if instance.assigned_qa != validated_data["assigned_qa"]:
-            if validated_data.get("should_alert_qa"):
-                self.try_send_assignment_emails(instance, "qa", validated_data, user)
-            orgs = instance.scratchorg_set.active().filter(
-                org_type=SCRATCH_ORG_TYPES.QA
-            )
-            for org in orgs:
-                org.queue_delete(originating_user_id=originating_user_id)
+        self._handle_reassign(
+            "dev", instance, validated_data, user, originating_user_id
+        )
+        self._handle_reassign("qa", instance, validated_data, user, originating_user_id)
         validated_data.pop("should_alert_dev", None)
         validated_data.pop("should_alert_qa", None)
         return super().update(instance, validated_data)
 
+    def _handle_reassign(
+        self, type_, instance, validated_data, user, originating_user_id
+    ):
+        if (
+            getattr(instance, f"assigned_{type_}")
+            != validated_data[f"assigned_{type_}"]
+        ):
+            if validated_data.get(f"should_alert_{type_}"):
+                self.try_send_assignment_emails(instance, type_, validated_data, user)
+            org_type = {"dev": SCRATCH_ORG_TYPES.Dev, "qa": SCRATCH_ORG_TYPES.QA}[type_]
+            reassigned_org = False
+            # We want to consider soft-deleted orgs, too:
+            orgs = [
+                *instance.scratchorg_set.active().filter(org_type=org_type),
+                *instance.scratchorg_set.inactive().filter(org_type=org_type),
+            ]
+            for org in orgs:
+                new_user = self._valid_reassign(
+                    type_, org, validated_data[f"assigned_{type_}"]
+                )
+                valid_commit = org.latest_commit == (
+                    instance.commits[0] if instance.commits else instance.origin_sha
+                )
+                if new_user and valid_commit and not reassigned_org:
+                    org.queue_reassign(
+                        new_user=new_user, originating_user_id=originating_user_id
+                    )
+                    reassigned_org = True
+                else:
+                    org.delete(originating_user_id=originating_user_id)
```

Comment:
@wlonk Right now this is also deleting the org on SF (via `model_mixins`, which calls `queue_delete`), which means it will not be available to be reused later. I'm not sure when we _do_ want to delete a scratch org on SF, vs when we do _not_ want to do that. I think we maybe want to distinguish between an org being explicitly deleted (because the user said "delete this org") vs deleted as a side-effect. I *think* we could do that by removing the logic in `model_mixins` that automatically deletes scratch orgs on SF when they're soft-deleted in Metecho (because in `perform_destroy` we're directly calling `queue_delete` to delete it on SF)... but I'd want a second set of eyes on that.

Reply:
I'm so confused by the intended behavior on this, at this point. I have no idea anymore. Maybe we should talk about it more?

Reply:
@wlonk I don't think anything has changed since our earlier discussion (https://github.com/oddbird/Metecho/pull/173#discussion_r442368141). You already made this change, so that it's now calling `org.delete()` instead of `org.queue_delete` -- with the intention of *not* deleting the org on SF automatically when the task role is reassigned. I'm pointing out that it's not quite working as intended yet, because the soft-delete eventually ends up calling `scratch_org.queue_delete()` anyway, as part of the soft-deletion.

In fact, that's already a bug in the current codebase, even before this PR: When you DELETE an org, it triggers `queue_delete`, which then calls `delete`, which then calls `queue_delete` _again_.

Reply:
`ScratchOrg.delete` doesn't call `queue_delete` though? We need to talk about this with code in front of us.

Reply:
@wlonk It does, via https://github.com/oddbird/Metecho/blob/main/metecho/api/model_mixins.py#L186 and https://github.com/oddbird/Metecho/blob/main/metecho/api/model_mixins.py#L226

Would you prefer to jump on a call? I'm around.

Reply:
What kind of terrible design queues a delete inside a `notify_`! This is why I need more code review; I think that's terrible, and I wrote it.